### PR TITLE
Struct support in gson-support

### PR DESCRIPTION
**Pull request by oldergod**
I will DRY the tests later, we need java support for Struct as well to do so.

### CODE REVIEW

File: wire-library/wire-gson-support/src/main/java/com/squareup/wire/StructTypeAdapter.kt [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        }
+        out.endArray()
+      }
+      else -> throw IllegalArgumentException("unexpected struct value: $any")
+    }
+    out.serializeNulls = serializeNulls
+  }
+
+  override fun read(input: JsonReader): Any {
+    // TODO(benoit) Ask for help. We kind of don't need it.
```
Comment (oldergod) [inline wire-library/wire-gson-support/src/main/java/com/squareup/wire/StructTypeAdapter.kt:57]: 👁️ 👁️  @swankjesse

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/StructTest.kt [hunk 2]
```diff
@@ -336,7 +336,113 @@ class StructTest {
...
+
+    val gson = GsonBuilder()
+        .registerTypeAdapterFactory(WireTypeAdapterFactory())
+        .disableHtmlEscaping()
+        .create()
+    assertJsonEquals(gson.toJson(wireAllStruct), json)
+    assertThat(gson.fromJson(json, AllStructs::class.java)).isEqualTo(wireAllStruct)
+  }
+
+  @Test fun structJsonRoundTripWithEmptyOrNestedMapAndListGson() {
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/StructTest.kt:382]: possible to parameterize these tests so they only need to be written once?

File: wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/StructTest.kt [hunk 3]
```diff
@@ -336,7 +336,113 @@ class StructTest {
...
+        value_e = mapOf("a" to 1.0),
+        value_f = listOf("a", 3.0)
+    )
+
+    val gson = GsonBuilder()
+        .registerTypeAdapterFactory(WireTypeAdapterFactory())
+        .disableHtmlEscaping()
+        .create()
+    assertJsonEquals(gson.toJson(wireAllStruct), json)
+    assertThat(gson.fromJson(json, AllStructs::class.java)).isEqualTo(wireAllStruct)
```
Comment (swankjesse) [inline wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/StructTest.kt:379]: I’m super curious what Gson does to track the internals here.

Comment (swankjesse) [reply wire-protoc-compatibility-tests/src/test/java/com/squareup/wire/StructTest.kt:379]: Might just be dispatching to `Object`.

**Pull request closed without merging**